File: api/langs/csharp/plump_message/PublicState.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Numerics;

namespace plump_message {

    public sealed class PublicState: IEquatable<PublicState>, ICloneable {
        public Serde.Option<ulong> guess;
        public ulong wins;
        public ulong score;

        public PublicState(Serde.Option<ulong> _guess, ulong _wins, ulong _score) {
            guess = _guess;
            wins = _wins;
            score = _score;
        }

        public void Serialize(Serde.ISerializer serializer) {
            serializer.increase_container_depth();
            TraitHelpers.serialize_option_u64(guess, serializer);
            serializer.serialize_u64(wins);
            serializer.serialize_u64(score);
            serializer.decrease_container_depth();
        }

        public int BincodeSerialize(byte[] outputBuffer) => BincodeSerialize(new ArraySegment<byte>(outputBuffer));

        public int BincodeSerialize(ArraySegment<byte> outputBuffer) {
            Serde.ISerializer serializer = new Bincode.BincodeSerializer(outputBuffer);
            Serialize(serializer);
            return serializer.get_buffer_offset();
        }

        public byte[] BincodeSerialize()  {
            Serde.ISerializer serializer = new Bincode.BincodeSerializer();
            Serialize(serializer);
            return serializer.get_bytes();
        }

        public static PublicState Deserialize(Serde.IDeserializer deserializer) {
            deserializer.increase_container_depth();
            PublicState obj = new PublicState(
            	TraitHelpers.deserialize_option_u64(deserializer),
            	deserializer.deserialize_u64(),
            	deserializer.deserialize_u64());
            deserializer.decrease_container_depth();
            return obj;
        }

        public static PublicState BincodeDeserialize(byte[] input) => BincodeDeserialize(new ArraySegment<byte>(input));

        public static PublicState BincodeDeserialize(ArraySegment<byte> input) {
            if (input == null) {
                 throw new Serde.DeserializationException("Cannot deserialize null array");
            }
            Serde.IDeserializer deserializer = new Bincode.BincodeDeserializer(input);
            PublicState value = Deserialize(deserializer);
            if (deserializer.get_buffer_offset() < input.Count) {
                 throw new Serde.DeserializationException("Some input bytes were not read");
            }
            return value;
        }
        public override bool Equals(object obj) => obj is PublicState other && Equals(other);

        public static bool operator ==(PublicState left, PublicState right) => Equals(left, right);

        public static bool operator !=(PublicState left, PublicState right) => !Equals(left, right);

        public bool Equals(PublicState other) {
            if (other == null) return false;
            if (ReferenceEquals(this, other)) return true;
            if (!guess.Equals(other.guess)) return false;
            if (!wins.Equals(other.wins)) return false;
            if (!score.Equals(other.score)) return false;
            return true;
        }

        public override int GetHashCode() {
            unchecked {
                int value = 7;
                value = 31 * value + guess.GetHashCode();
                value = 31 * value + wins.GetHashCode();
                value = 31 * value + score.GetHashCode();
                return value;
            }
        }

        /// <summary>Creates a shallow clone of the object.</summary>
        public PublicState Clone() => (PublicState)MemberwiseClone();

        object ICloneable.Clone() => Clone();

    }

} // end of namespace plump_message

File: api/langs/java/plump_message/PublicState.java
package plump_message;


public final class PublicState {
    public final java.util.Optional<@com.novi.serde.Unsigned Long> guess;
    public final @com.novi.serde.Unsigned Long wins;
    public final @com.novi.serde.Unsigned Long score;

    public PublicState(java.util.Optional<@com.novi.serde.Unsigned Long> guess, @com.novi.serde.Unsigned Long wins, @com.novi.serde.Unsigned Long score) {
        java.util.Objects.requireNonNull(guess, "guess must not be null");
        java.util.Objects.requireNonNull(wins, "wins must not be null");
        java.util.Objects.requireNonNull(score, "score must not be null");
        this.guess = guess;
        this.wins = wins;
        this.score = score;
    }

    public void serialize(com.novi.serde.Serializer serializer) throws com.novi.serde.SerializationError {
        serializer.increase_container_depth();
        TraitHelpers.serialize_option_u64(guess, serializer);
        serializer.serialize_u64(wins);
        serializer.serialize_u64(score);
        serializer.decrease_container_depth();
    }

    public byte[] bincodeSerialize() throws com.novi.serde.SerializationError {
        com.novi.serde.Serializer serializer = new com.novi.bincode.BincodeSerializer();
        serialize(serializer);
        return serializer.get_bytes();
    }

    public static PublicState deserialize(com.novi.serde.Deserializer deserializer) throws com.novi.serde.DeserializationError {
        deserializer.increase_container_depth();
        Builder builder = new Builder();
        builder.guess = TraitHelpers.deserialize_option_u64(deserializer);
        builder.wins = deserializer.deserialize_u64();
        builder.score = deserializer.deserialize_u64();
        deserializer.decrease_container_depth();
        return builder.build();
    }

    public static PublicState bincodeDeserialize(byte[] input) throws com.novi.serde.DeserializationError {
        if (input == null) {
             throw new com.novi.serde.DeserializationError("Cannot deserialize null array");
        }
        com.novi.serde.Deserializer deserializer = new com.novi.bincode.BincodeDeserializer(input);
        PublicState value = deserialize(deserializer);
        if (deserializer.get_buffer_offset() < input.length) {
             throw new com.novi.serde.DeserializationError("Some input bytes were not read");
        }
        return value;
    }

    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        PublicState other = (PublicState) obj;
        if (!java.util.Objects.equals(this.guess, other.guess)) { return false; }
        if (!java.util.Objects.equals(this.wins, other.wins)) { return false; }
        if (!java.util.Objects.equals(this.score, other.score)) { return false; }
        return true;
    }

    public int hashCode() {
        int value = 7;
        value = 31 * value + (this.guess != null ? this.guess.hashCode() : 0);
        value = 31 * value + (this.wins != null ? this.wins.hashCode() : 0);
        value = 31 * value + (this.score != null ? this.score.hashCode() : 0);
        return value;
    }

    public static final class Builder {
        public java.util.Optional<@com.novi.serde.Unsigned Long> guess;
        public @com.novi.serde.Unsigned Long wins;
        public @com.novi.serde.Unsigned Long score;

        public PublicState build() {
            return new PublicState(
                guess,
                wins,
                score
            );
        }
    }
}
